Thread toolchain through to error message

kinnison opened the pull request:
In the component-unavailable error case, we need to report the
full toolchain rather than just the component and target name.
This threads it through and reports it in the error message.

Fixes: #1596

This is not ideal right now, but it's a starting point for discussion.

Currently the error message ends up looking something like:

```
error: component 'rust-std' for target 'x86_64-unknown-linux-gnu' is unavailable for download for toolchain 'nightly-x86_64-unknown-linux-gnu'
```

Clearly the repetition of the target name in the toolchain name isn't ideal, I've been considering
reformatting the message but didn't want to do so without some further input on what might be better.

h-michael commented:
How about `download for 'nightly' channel`?

kinnison commented:
I need to work out the right way to produce a "short" toolchain, but yes that wording migh be nicer @h-michael

kinnison commented:
I've managed to make it say that (and if there's a date associated, channel-date) so I'd appreciate reviews / opinions.  If noone has anything negative to say in the next day or so I'll un-WIP this.

nrc commented:
Looks good. We would want to fix the repetition. We should probably split the toolchain into channel and host (is it host or target, actually?) and only report the channel and host if it is different to the target

kinnison commented:
So I changed the toolchain to just be `channel` or `channel-date` but I suppose we properly need to distinguish the toolchain host *if* it doesn't match the host of the system running rustup.  The goal would be to ensure that the toolchain statement in the error message should match what is being used with the minimal content possible.

I think this probably needs careful thought to decide what's ideal.  What is here now is better, but perhaps not ideal.

dwijnand commented:
Removing the repetition seems to me like a nice-to-have of lesser value than resolving #1596, so I'm happy to accept this as is and trim it later/never.

bors commented:
:umbrella: The latest upstream changes (presumably #1643) made this pull request unmergeable. Please resolve the merge conflicts.

kinnison committed:
Thread toolchain through to error message

In the component-unavailable error case, we need to report the
full toolchain rather than just the component and target name.
This threads it through and reports it in the error message.

Fixes: #1596

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

kinnison commented:
I've opted to rebase this onto mater and as per @dwijnand 's observation that something is better than nothing, I commend it for review and ideally merge.

I'm not sure entirely how I'd make the change which we've identified as most desirable so I suggest that we take this for now, and consider fixing up the suggestions further after simplification has proceeded somewhat.
